crates/ecs/src/filter.rs

```diff
@@ -0,0 +1,345 @@
+            type BorrowedData<'components> = (
+                <L as SystemParameter>::BorrowedData<'components>,
+                <R as SystemParameter>::BorrowedData<'components>,
+            );
+
+            fn borrow(world: &World) -> Self::BorrowedData<'_> {
+                (
+                    <L as SystemParameter>::borrow(world),
+                    <R as SystemParameter>::borrow(world),
+                )
+            }
+
+            unsafe fn fetch_parameter(
+                borrowed: &mut Self::BorrowedData<'_>,
+                entity: Entity,
+            ) -> Option<Self> {
+                let (left_borrow, right_borrow) = borrowed;
+                let left = <L as SystemParameter>::fetch_parameter(left_borrow, entity);
+                let right = <R as SystemParameter>::fetch_parameter(right_borrow, entity);
+                (left.is_some() $op right.is_some()).then(|| Self {
+                    left: PhantomData::default(),
+                    right: PhantomData::default(),
+                })
+            }
+
+            fn component_accesses() -> Vec<ComponentAccessDescriptor> {
+                [
+                    <L as SystemParameter>::component_accesses(),
+                    <R as SystemParameter>::component_accesses(),
+                ]
+                .concat()
+            }
+        }
+    }
+}
+
+binary_filter_operation!(And, &&, "And", "and");
+binary_filter_operation!(Or, ||, "Or", "or");
+binary_filter_operation!(Xor, ^, "Xor", "xor");
+
+/// A query filter that inverts another filter.
+///
+/// # Example
+/// ```
+/// # use ecs::filter::{Not, With};
+/// # use ecs::Read;
+/// # #[derive(Debug)]
+/// # struct Position;
+/// # struct Player;
+/// fn non_player_position(position: Read<Position>, _: Not<With<Player>>) {
+///     println!("A non-player entity is at position {:?}.", position);
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Not<T: SystemParameter> {
+    phantom: PhantomData<T>,
+}
+
+impl<T: SystemParameter> SystemParameter for Not<T> {
+    type BorrowedData<'components> = <T as SystemParameter>::BorrowedData<'components>;
+
+    fn borrow(world: &World) -> Self::BorrowedData<'_> {
+        <T as SystemParameter>::borrow(world)
+    }
+
+    unsafe fn fetch_parameter(
+        borrowed: &mut Self::BorrowedData<'_>,
+        entity: Entity,
+    ) -> Option<Self> {
+        if <T as SystemParameter>::fetch_parameter(borrowed, entity).is_some() {
+            None
+        } else {
+            Some(Self {
+                phantom: PhantomData::default(),
+            })
+        }
+    }
+
+    fn component_accesses() -> Vec<ComponentAccessDescriptor> {
+        <T as SystemParameter>::component_accesses()
+    }
+}
+
+/// A query filter that matches any entity.
+///
+/// # Example
+/// ```
+/// # use ecs::filter::Any;
+/// fn all_entities(_: Any) {
+///     println!("Hello from an entity!");
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Any {}
```

Comment:
> From an end-users perspective, the Any filter has very few places where it is useful. It is more useful when writing test for different logical identities.

I see, could you perhaps move this filter into the test module then? Since it seems a bit redundant for ordinary non-test use, as you say. We can always move it back out if it turns out to have a use.

> `WithEntity` is possible to implement and would be more useful, but it should not be considered a filter because it is returning data.

Makes sense! 

> I suggest that we just implement SystemParameter for Entity for this purpose instead, and then Any will be the filter equivalence to Enity in the same way as With is the filter equivalence to Read. Then, it should be possible implement a system like this: ...

Great, that looks perfect. (Though, you don't need to implement that in this PR -- we'll save that for when it actually has a use.)